# Public data resources

Public datasets are a valuable resource for bioinformatics projects. They can be used to test new algorithms, develop new tools, and validate your results. Here are some public data resources that you may find useful.

The [Australian BioCommons Data Management Collection](https://australianbiocommons.github.io/Learning-Library/data_management) is designed to support life sciences researchers understand the best practices and procedures involved in the stages of the data lifecycle. The collection consists of training materials and documentation from reputable sources such as EMBL-EBI Training and NCBI Training, and covers topics like data sharing and data reuse.

## General next generation sequencing 

* Sequence Read Archive: https://trace.ncbi.nlm.nih.gov/Traces/index.html  
* European Nucleotide Archive: https://www.ebi.ac.uk/ena/browser/ 
* GenBank: https://www.ncbi.nlm.nih.gov/genbank/ 

## Genome browsers

* Ensembl Genome Browser: https://asia.ensembl.org/index.html ([FTP site](https://asia.ensembl.org/info/data/ftp/index.html))
* UCSC Genome Browser: https://genome.ucsc.edu/index.html
* NCBI Genome Data Viewer: https://www.ncbi.nlm.nih.gov/gdv/

## Genomics 

* 1000 Genomes Project: https://www.internationalgenome.org/
* dbSNP: https://www.ncbi.nlm.nih.gov/snp/
* GnomAD: https://gnomad.broadinstitute.org/
* GWAS Catalog: https://www.ebi.ac.uk/gwas/

## Transcriptomics

* Gene Expression Omnibus: https://www.ncbi.nlm.nih.gov/geo/
* ArrayExpress: https://www.ebi.ac.uk/arrayexpress/
* GTEx Portal: https://www.gtexportal.org/home/

## Proteomics 

* PRIDE: https://www.ebi.ac.uk/pride/archive/
* PeptideAtlas: http://www.peptideatlas.org/
* SWATHAtlas: https://db.systemsbiology.net/sbeams/cgi/PeptideAtlas/SwathAtlas
* UniProt: https://www.uniprot.org/
* InterPro: https://www.ebi.ac.uk/interpro/

## Metabolomics

* Metabolghts: https://www.ebi.ac.uk/metabolights/
* Metabolomics Workbench: https://www.metabolomicsworkbench.org/
* Human Metabolome Database: http://www.hmdb.ca/

## Single cell 

* Human Cell Atlas: https://www.humancellatlas.org/
* scRNAS-seqDB: https://bioinfo.uth.edu/scrnaseqdb/
* Cell Ranger Public Data: https://support.10xgenomics.com/single-cell-gene-expression/datasets

## Microbiome

* MG-RAST: https://www.mg-rast.org/
* QIITA: https://qiita.ucsd.edu/
* Greengenes: http://greengenes.lbl.gov/

## Structural biology (proteomics, genomics, transcriptomics)

* Protein Data Bank: https://www.rcsb.org/
* TADKB: http://tad.bioinfo.sdu.edu.cn/
* 4D nucleosome: http://www.4dnucleome.org/
* Genome Structure Browser: https://gsdb.mu.hekademeia.org/

## Epigenetics 

* ENCODE: https://www.encodeproject.org/
* MethBank: http://bigd.big.ac.cn/methbank/index
* Roadmap Epigenomics: https://www.roadmapepigenomics.org/

## Clinical 

* OMIM: https://www.omim.org/
* HapMap: https://www.ncbi.nlm.nih.gov/projects/SNP/snp_viewTable.cgi?handle=hapmap
* The Cancer Genome Atlas Project: https://www.cancer.gov/ccg/research/genome-sequencing/tcga
* CADD: https://cadd.gs.washington.edu/
* ClinVar: https://www.ncbi.nlm.nih.gov/clinvar/
* Genome In A Bottle: https://www.nist.gov/programs-projects/genome-bottle
* COSMIC: https://cancer.sanger.ac.uk/cosmic

## Animal 

* OMIA: https://omia.org/home/
* Animal QTLdb: https://www.animalgenome.org/QTLdb/
* Mouse Genome Informatics: http://www.informatics.jax.org/

## Microbes 

* Kraken2/Bracken: https://benlangmead.github.io/aws-indexes/k2 
* PATRIC: https://www.patricbrc.org/
* IMG/M: https://img.jgi.doe.gov/cgi-bin/m/main.cgi

## Plant

* Phytozome: https://phytozome-next.jgi.doe.gov/
* PlantGDB: http://www.plantgdb.org/

## Gene ontologies and annotations

* Gene Ontology: http://geneontology.org/
* EDAM ontology: https://edamontology.org/
* OLS: https://www.ebi.ac.uk/ols/index
* Human phenotype ontology: https://hpo.jax.org/app/
* MEROPS: https://www.ebi.ac.uk/merops/ 
* KEGG: https://www.genome.jp/kegg/
