# Bioinformatics file formats explained 

A more in depth explainer of some file formats is provided by [UCONN Computational Biology Core](https://bioinformatics.uconn.edu/resources-and-events/tutorials-2/file-formats-tutorial/).

## FASTA

### What is it for?

FASTA is a text-based format for representing either nucleotide sequences or peptide sequences, in which nucleotides or amino acids are represented using single-letter codes. Sequences are typically represented in the standard [IUPAC](http://www.chem.qmul.ac.uk/iupac/AminoAcid) nucleotide or amino acid codes.

It is routinely used for sequence representation (e.g. reference genomes and transcriptomes), database searches, and multiple sequence alignments.  

### How is it formatted?

* File extensions:`.fa`, `.fasta`, `.fsa`

A sequence in FASTA format begins with a single line description, distinguished from the actual sequence data by a greater-than (">") symbol. The description line is followed by lines of sequence data. The number of characters per line of sequence is generally limited to 80 characters. 

[Details on FASTA formatting](https://www.ncbi.nlm.nih.gov/genbank/fastaformat/)

```
>MyGene                 
CCTCTCGGAGCTGGAAATGCAGCTATTGAGATCTTCGAATGCTGCGGAGCTGGAGGCGGA
GGCAGCTGGGGAGGTCCGAGCGATGTGACCAGGCCGCCATCGCTCGTCTCTTCCTCTCTC
CTGCCGCCTCCTGTGTCGAAAATAACTTTTTTAGTCTAAAGAAAGAAAGACAAAAGTAGT
CGTCCGCCCCTCACGCCCTCTCTTCCTCTCAGCCTTCCGCCCGGTGAGGAAGCCCGGGGT
GGCTGCTCCGCCGTCGGGGCCGCGCCGCCGAGCCCCAGCGCCCCGGGCCGCCCCCGCACG
CCGCCCCCATGCATCCCTTCTACACCCGGGCCGCCACCATGATAGGCGAGATCGCCGCCG
CCGTGTCCTTCATCTCCAAGTTTCTCCGCACCAAGGGGCTGACGAGCGAGCGACAGCTGC
AGACCTTCAGCCAGAGCCTGCAGGAGCTGCTGGCAGAACATTATAAACATCACTGGTTCC
CAGAAAAGCCATGCAAGGGATCGGGTTACCGTTGTATTCGCATCAACCATAAAATGGATC
```
```
>MyProtein
MAVTQTAQACDLVIFGAKGDLARRKLLPSLYQLEKAGQLNPDTRIIGVGRADWDKAAYTK
VVREALETFMKETIDEGLWDTLSARLDFCNLDVNDTAAFSRLGAMLDQKNRITINYFAMP
PSTFGAICKGLGEAKLNAKPARVVMEKPLGTSLATSQEINDQVGEYFEECQVYRIDHYLG
KETVLNLLALRFANSLFVNNWDNRTIDHVEITV 
```

## FASTQ

### What is it for?

FASTQ format is a text-based format for storing both a biological sequence and its corresponding quality scores. It is commonly used to store high-throughput sequencing data, such as that generated by Illumina, Oxford Nanopore, and PacBio sequencers. It was developed by the Sanger Institute to group together sequence and corresponding quality scores ([Phred scores](https://en.wikipedia.org/wiki/Phred_quality_score)) in a single file.

### How is it formatted?

* File extensions:`.fq`, `.fastq`, `.fq.gz`, `.fastq.gz`

Each FASTQ entry is associated with 4 lines: 
* Line 1: Begins with a '@' character and is followed by a sequence identifier and optional descriptions (e.g. instrument name, flowcell id).
* Line 2: The sequence data represented as [IUPAC](http://www.chem.qmul.ac.uk/iupac/AminoAcid) codes.
* Line 3: Begins with a '+' character and is optionally followed by the same sequence identifier and description as line 1.
* Line 4: The quality scores for the sequence data. The quality scores are encoded using ASCII characters, with each character representing a Phred quality score.

FASTQ files are often compressed in gzipped format to reduce space.  


[Details on FASTQ formatting](https://en.wikipedia.org/wiki/FASTQ_format)

```
@K00188:208:HFLNGBBXX:3:1101:1428:1508 2:N:0:CTTGTA
ATAATAGGATCCCTTTTCCTGGAGCTGCCTTTAGGTAATGTAGTATCTNATNGACTGNCNCCANANGGCTAAAGT
+
AAAFFJJJJJJJJJJJJJJJJJFJJFJJJJJFJJJJJJJJJJJJJJJJ#FJ#JJJJF#F#FJJ#F#JJJFJJJJJ
```

## BAM

### What is it for?

BAM (Binary Alignment/Map) is a binary file format used to store sequence alignment data. It is a compressed, indexed version of the Sequence Alignment/Map (SAM) format. It is not human readable, you will need to use software like [samtools](http://www.htslib.org/) to view their contents. BAM files are commonly used to store the results of sequence alignment, such as the alignment of short reads to a reference genome.

### How is it formatted?

* File extension:`.bam`

BAM files are binary files that store the following information:
* Header: Contains metadata about the alignment, such as the reference genome used, the alignment algorithm, and the read group information.
* Alignment records: Each record contains information about a single read alignment, such as the read name, the reference sequence name, the alignment position, and the quality scores.
* 

BAM files are indexed to allow for efficient random access to the alignment records. Their indexing allows for rapid retrieval of reads aligned to a specific region of the reference genome. Indexes are stored as additional files with the `.bai` extension.

[Details on BAM formatting](https://support.illumina.com/help/BS_App_RNASeq_Alignment_OLH_1000000006112/Content/Source/Informatics/BAM-Format.htm)

```
@HD	VN:1.6	SO:coordinate
@SQ	SN:chr1	LN:248956422
@RG	ID:UM0098	SM:Sample1
1	113	chr1	497	37	37M	=	100338662	0	CGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG	0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>	XT:A:U	NM:i:0	SM:i:37	AM:i:0	X0:i:1	X1:i:0	XM:i:0	XO:i:0	XG:i:0	MD:Z:37
19	99	chr1	17644	0	37M	=	17919	314	TATGACTGCTAATAATACCTACACATGTTAGAACCAT	>>>>>>>>>>>>>>>>>>>><<>>><<>>4::>>:<9	RG:Z:UM0098	XT:A:R	NM:i:0	SM:i:0	AM:i:0	X0:i:4	X1:i:0	XM:i:0	XO:i:0	XG:i:0	MD:Z:37
19	147	chr1	17919	0	18M2D19M	=	17644	-314	GTAGTACCAACTGTAAGTCCTTATCTTCATACTTTGT	;44999;499<8<8<<<8<<><<<<><7<;<<<>><<	XT:A:R	NM:i:2	SM:i:0	AM:i:0	X0:i:4	X1:i:0	XM:i:0	XO:i:1	XG:i:2	MD:Z:18^CA19
9	83	chr1	21678	0	8M2I27M	=	21469	-244	CACCACATCACATATACCAAGCCTGGCTGTGTCTTCT	<;9<<5><<<<><<<>><<><>><9>><>>>9>>><>	XT:A:R	NM:i:2	SM:i:0	AM:i:0	X0:i:5	X1:i:0	XM:i:0	XO:i:1	XG:i:2	MD:Z:35
```

## SAM

### What is it for?

SAM (Sequence Alignment/Map) is a text-based file format used to store sequence alignment data. It is the uncompressed version of the BAM format. It is human readable. SAM files are commonly used to store the results of sequence alignment, such as the alignment of short reads to a reference genome.

### How is it formatted?

* File extension: `.sam`

SAM files are text files that store the following information:
* Header: Contains metadata about the alignment, such as the reference genome used, the alignment algorithm, and the read group information.
* Alignment records: Each record contains information about a single read alignment, such as the read name, the reference sequence name, the alignment position, and the quality scores.
* Optional fields: Additional information about the alignment, such as the mapping quality, the number of mismatches, and the alignment score.

[Details on SAM formatting](https://samtools.github.io/hts-specs/SAMv1.pdf)

```
@HD	VN:1.6	SO:coordinate
@SQ	SN:chr1	LN:248956422
@RG	ID:UM0098	SM:Sample1
1	113	chr1	497	37	37M	=	100338662	0	CGGGTCTGACCTGAGGAGAACTGTGCTCCGCCTTCAG	0;==-==9;>>>>>=>>>>>>>>>>>=>>>>>>>>>>	XT:A:U	NM:i:0	SM:i:37	AM:i:0	X0:i:1	X1:i:0	XM:i:0	XO:i:0	XG:i:0	MD:Z:37
19	99	chr1	17644	0	37M	=	17919	314	TATGACTGCTAATAATACCTACACATGTTAGAACCAT	>>>>>>>>>>>>>>>>>>>><<>>><<>>4::>>:<9	RG:Z:UM0098	XT:A:R	NM:i:0	SM:i:0	AM:i:0	X0:i:4	X1:i:0	XM:i:0	XO:i:0	XG:i:0	MD:Z:37
19	147	chr1	17919	0	18M2D19M	=	17644	-314	GTAGTACCAACTGTAAGTCCTTATCTTCATACTTTGT	;44999;499<8<8<<<8<<><<<<><7<;<<<>><<	XT:A:R	NM:i:2	SM:i:0	AM:i:0	X0:i:4	X1:i:0	XM:i:0	XO:i:1	XG:i:2	MD:Z:18^CA19
9	83	chr1	21678	0	8M2I27M	=	21469	-244	CACCACATCACATATACCAAGCCTGGCTGTGTCTTCT	<;9<<5><<<<><<<>><<><>><9>><>>>9>>><>	XT:A:R	NM:i:2	SM:i:0	AM:i:0	X0:i:5	X1:i:0	XM:i:0	XO:i:1	XG:i:2	MD:Z:35
```

## CRAM

### What is it for?

CRAM (Compressed Read Alignment Map) is a compressed file format for storing sequence alignment data. It was developed to reduce the storage requirements of large genomic datasets by using reference-based compression. CRAM files can achieve significant reductions in file size compared to BAM files while preserving all the essential information required for downstream analysis. 

### How is it formatted?

* File extension: `.cram`

CRAM files use reference-based compression, where the sequence reads are stored as differences relative to a reference genome. This allows for a higher compression ratio compared to BAM. The format includes the following components:
* Header: Similar to BAM, contains metadata about the alignment, including reference sequences and alignment parameters.
* Data blocks: Contain compressed sequence alignment data, encoded in a way that exploits redundancy within the data and the reference genome.

To view CRAM files, tools like [samtools](http://www.htslib.org/) are required. CRAM files can also be indexed for fast random access, with indexes stored in `.crai` files.

[Details on CRAM formatting](https://en.wikipedia.org/wiki/CRAM_(file_format))

## VCF

### What is it for?

VCF (Variant Call Format) is a text file format used for storing genetic variation data, such as single nucleotide polymorphisms (SNPs), insertions, deletions, and structural variants. VCF files are widely used in genomics to represent the results of variant calling analyses. They contain information about the genomic position of the variant, the reference and alternate alleles, and quality scores associated with the variant call.

### How is it formatted?


* File extension: `.vcf`

VCF files consist of a header section and a data section:
* Header: Begins with `##` and contains metadata lines that describe the file format, reference genome, and other relevant information. The header is followed by a line starting with `#` that defines the column labels for the data section.
* Data: Each line represents a single variant and includes the following fields:
  * CHROM: Chromosome name
  * POS: Position of the variant
  * ID: Identifier for the variant
  * REF: Reference base(s)
  * ALT: Alternate base(s)
  * QUAL: Quality score for the variant call
  * FILTER: Filter status
  * INFO: Additional information about the variant
  * FORMAT: (Optional) Format of the genotype data
  * Sample columns: (Optional) Genotype data for each sample

[Details on VCF formatting](https://samtools.github.io/hts-specs/VCFv4.3.pdf)

```
##fileformat=VCFv4.3
##reference=GRCh38
#CHROM POS ID REF ALT QUAL FILTER INFO FORMAT Sample1 Sample2
1  879317 .  G  A  29.1  PASS  DP=14;AF=0.5  GT:GQ  0/1:48  0/0:99
1  879602 rs6054257 G  A  14.7  q10  DP=11;AF=0.017  GT:AD:DP:GQ:PL  0/0:10,1:11:43:0,43,255  0/0:12,0:12:37:0,37,255
```

## GTF

### What is it for?

GTF (Gene Transfer Format) is a text file format used for storing information about gene structure, such as exon, intron, and coding sequences. It is commonly used in genomics to represent gene annotations.

### How is it formatted?

* File extension: `.gtf`

GTF files are tab-delimited and consist of nine columns:
1. **Seqname:** Name of the sequence (chromosome or scaffold)
2. **Source:** Source of the annotation (e.g., a prediction algorithm)
3. **Feature:** Type of feature (e.g., exon, CDS, gene)
4. **Start:** Start position of the feature
5. **End:** End position of the feature
6. **Score:** Score of the feature (e.g., confidence score)
7. **Strand:** Strand of the feature (+ or -)
8. **Frame:** Reading frame of the feature (0, 1, or 2)
9. **Attribute:** Additional information about the feature in a key-value format

[Details on GTF formatting](https://genome.ucsc.edu/FAQ/FAQformat.html#format4)

```text
chr1  HAVANA  gene  11869  14409  .  +  .  gene_id "ENSG00000223972"; gene_name "DDX11L1";
chr1  HAVANA  transcript  11869  14409  .  +  .  gene_id "ENSG00000223972"; transcript_id "ENST00000456328"; gene_name "DDX11L1";
chr1  HAVANA  exon  11869  12227  .  +  .  gene_id "ENSG00000223972"; transcript_id "ENST00000456328"; exon_number "1"; gene_name "DDX11L1";
chr1  HAVANA  exon  12613  12721  .  +  .  gene_id "ENSG00000223972"; transcript_id "ENST00000456328"; exon_number "2"; gene_name "DDX11L1";
```
## GFF

### What is it for?

GFF (General Feature Format) is a text file format used for describing genes and other features of DNA, RNA, and protein sequences. It is similar to GTF but has a simpler structure and is more flexible in terms of the types of annotations it can represent.

### How is it formatted?

* File extension: `.gff`, `.gff3`


GFF files are tab-delimited and consist of nine columns:
1. **Seqname:** Name of the sequence (chromosome or scaffold)
2. **Source:** Source of the annotation (e.g., a prediction algorithm)
3. **Feature:** Type of feature (e.g., exon, CDS, gene)
4. **Start:** Start position of the feature
5. **End:** End position of the feature
6. **Score:** Score of the feature (e.g., confidence score)
7. **Strand:** Strand of the feature (+ or -)
8. **Frame:** Reading frame of the feature (0, 1, or 2)
9. **Attribute:** Additional information about the feature in a key-value format

GFF3 is the most recent version and includes additional features like hierarchical relationships between features.

[Details on GFF formatting](https://genome.ucsc.edu/FAQ/FAQformat.html#format3)

```text
##gff-version 3
chr1  .  gene  1300  9000  .  +  .  ID=gene00001;Name=EDEN
chr1  .  mRNA  1300  9000  .  +  .  ID=mRNA00001;Parent=gene00001;Name=EDEN.1
chr1  .  exon  1300  1500  .  +  .  ID=exon00001;Parent=mRNA00001
chr1  .  exon  1050  9000  .  +  .  ID=exon00002;Parent=mRNA00001
chr1  .  CDS  1300  1500  .  +  0  ID=cds00001;Parent=mRNA00001
chr1  .  CDS  3000  3902  .  +  2  ID=cds00002;Parent=mRNA00001
chr1  .  CDS  5000  9000  .  +  0  ID=cds00003;Parent=mRNA00001
```