# Learning materials 

Bioinformatics is an interdisciplinary field that leverages computational tools and techniques to analyse and interpret biological data. This resources page provides foundational knowledge and tools essential for bioinformaticians, focusing on key areas such as the Linux shell, code management, R and Python programming languages, and Git for version control. For those who are new to the dry lab (the computer), you'll need to learn some basics of the shell, a coding language or 2, and code management practices. The resources below can guide you through these topics.

* [What is bioinformatics?](#what-is-bioinformatics)
* [The linux shell](#the-linux-shell)
* [Intro to HPC](#into-to-hpc)
* [Code management](#code-management)
* [Software management](#software-management)
* [R and Python](#r-and-python)
* [Varios 'omics techniques](#various-omics-techniques)

## What is bioinformatics?

So many different types of data and applications. These resources can help you contextualise all the jargon and understand the practicalities of doing bioinformatics and computational biology. 

* https://book.biologistsguide2computing.com/en/stable/ 
* https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008645 
* https://www.nature.com/articles/nbt.2740 
* https://web.stanford.edu/class/bios221/book/ 
* https://github.com/applied-bioinformatics/iab2 
* https://genohub.com/next-generation-sequencing-handbook/ 
* https://medium.com/@gearthdexter/software-tooling-bioinformatics-introduction-ebcb9d020f95 

## The linux shell 

The Linux shell is a powerful interface for interacting with your computer's operating system or the HPC system you may be working on. It allows you to run commands, manage files, and automate tasks, making it an essential tool for bioinformaticians. We recommend the following materials to help you learn the basics:

* https://linuxjourney.com/
* https://sandbox.bio/
* https://github.com/hbctraining/Intro-to-shell-flipped
* https://swcarpentry.github.io/shell-novice/
* http://andrewjrobinson.github.io/training_docs/tutorials/unix/
* https://www.edx.org/learn/linux/the-linux-foundation-introduction-to-linux 

## Into to HPC

High performance computing (HPC) systems are used to run computationally intensive analyses. Chances are if you're working with big genomes or large sample cohorts, you'll need to use an HPC. Keep in mind that depending on the system, you may be working with a different job scheduler (e.g. PBSpro vs SLURM), so your code might not cleanly port over from one system to another. If you are new to HPC and working on Australian HPC systems, you may want to start with the following courses and resources:

* Pro tips for scaling bioinformatics to HPC: https://www.biocommons.org.au/events/workflows-hpc
* [Ten simple rules for large-scale data processing](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1009757)
* Intro to Artemis HPC: https://sydney-informatics-hub.github.io/training.artemis/ 
* Intro to Gadi HPC: https://sydney-informatics-hub.github.io/training.gadi.intro/ 
* Intro to Gadi workshop calendar: https://opus.nci.org.au/display/Help/Introduction+to+Gadi 
* Pawsey training materials: https://pawseysc.github.io/ 
* Carprentries HPC: https://www.hpc-carpentry.org/index.html 
* http://andrewjrobinson.github.io/training_docs/tutorials/hpc/

## Code management

Code management is crucial for maintaining and organising your scripts and software. It ensures reproducibility and collaboration in bioinformatics projects. Managing your code can save you time and effort in the long run. It is increasingly an expectation of publications that you share your code. We recommend the following resources to help you learn the basics of code management:

* https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1000424 
* https://sydney-informatics-hub.github.io/intro-git/
* https://swcarpentry.github.io/git-novice/
* https://skills.github.com/ 

## Software management

Bioinformatics software typically have a lot of dependencies. Managing software dependencies is essential for ensuring that your code runs smoothly on different systems. Software management tools like containers and conda and workflow management tools like Nextflow can help you manage software dependencies and create reproducible environments. We recommend the following resources to help you learn the basics of software management:

* https://www.melbournebioinformatics.org.au/tutorials/tutorials/docker/docker/ 
* https://www.melbournebioinformatics.org.au/tutorials/tutorials/galaxy-workflows/galaxy-workflows/ 
* https://www.blopig.com/blog/2021/05/singularity-a-guide-for-the-bewildered-bioinformatician/ 
* https://protocols.hostmicrobe.org/conda 
* https://astrobiomike.github.io/unix/conda-intro 

## R and Python 

### R 

* Flipped intro to R https://github.com/hbctraining/Intro-to-R-flipped
* Intro to R https://monashdatafluency.github.io/r-intro-2/
* R for genomics: https://datacarpentry.org/genomics-r-intro/ 
* Monash R workshop https://monashbioinformaticsplatform.github.io/r-more/ 
* Using Rmarkdown https://rmarkdown.rstudio.com/authoring_quick_tour.html
* WEHI's tidyverse intro: https://bookdown.org/ansellbr/WEHI_tidyR_course_book/
* Working in the Tidyverse: http://www.hiercourse.com/docs/Working_in_the_Tidyverse.pdf 
* Code School R: http://tryr.codeschool.com/ 
* Coursera R programming: https://www.coursera.org/course/rprog
* Choose your own adventure for stats tests: https://www.statstest.com/ 
* R graph gallery: https://www.r-graph-gallery.com/index.html 
* Datacamp quick R graphs: https://www.statmethods.net/graphs/index.html 


### Python 
* Python novice: https://swcarpentry.github.io/python-novice-inflammation/
* Python plotting: https://swcarpentry.github.io/python-novice-gapminder/
* Official Python tutorial: https://docs.python.org/3/tutorial/ 
* Python for everybody: https://www.py4e.com/lessons 
* Data processing: http://opentechschool.github.io/python-data-intro/ 

## Various 'omics techniques


